### PR TITLE
[mlir][memref] Add memref alias folding for masked transfers

### PR DESCRIPTION
**Pull request by qedawkins**
The contents of a mask on a masked transfer are unaffected by the particular region of memory being read/stored to, so just forward the mask in subview folding patterns.

### CODE REVIEW

File: mlir/test/Dialect/MemRef/fold-memref-alias-ops.mlir [hunk 1]
```diff
@@ -266,6 +266,63 @@ func.func @fold_vector_transfer_write_with_inner_rank_reduced_subview(
...
+//   CHECK-DAG: #[[MAP1:.+]] = affine_map<()[s0, s1] -> (s0 + s1)>
+//       CHECK: func @fold_masked_vector_transfer_read_with_subview
+//  CHECK-SAME:    %[[ARG0:[a-zA-Z0-9]+]]: memref<?x?xf32, strided<[?, ?], offset: ?>>
+//  CHECK-SAME:    %[[ARG1:[a-zA-Z0-9]+]]: index
+//  CHECK-SAME:    %[[ARG2:[a-zA-Z0-9]+]]: index
+//  CHECK-SAME:    %[[ARG3:[a-zA-Z0-9]+]]: index
+//  CHECK-SAME:    %[[ARG4:[a-zA-Z0-9]+]]: index
+//  CHECK-SAME:    %[[ARG5:[a-zA-Z0-9]+]]: index
+//  CHECK-SAME:    %[[ARG6:[a-zA-Z0-9]+]]: index
+//  CHECK-SAME:    %[[ARG7:[a-zA-Z0-9]+]]: vector<4xi1>
```
Comment (antiagainst) [inline mlir/test/Dialect/MemRef/fold-memref-alias-ops.mlir:290]: Nit: I typically prefer naming as a docmentation, e.g., `MASK` would be better than `ARG7` to me.

**Review by antiagainst — APPROVED** [review]

**Comment by antiagainst** [comment]
A tricky part with subview ops in general is rank reducing behavior. Thinking a bit actually it should be fine given that folding rank reducing subview ops into the original memref would just mean we see extra dims in the transfer op permuation map. That will be [compressed when inferTransferOpMaskType](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Vector/IR/VectorOps.cpp#L3758-L3773), so we can still use the same mask.

**Comment by nicolasvasilache** [comment]
> A tricky part with subview ops in general is rank reducing behavior. Thinking a bit actually it should be fine given that folding rank reducing subview ops into the original memref would just mean we see extra dims in the transfer op permuation map. That will be [compressed when inferTransferOpMaskType](https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/Vector/IR/VectorOps.cpp#L3758-L3773), so we can still use the same mask.

Let's add some rank-reducing tests please

File: mlir/test/Dialect/MemRef/fold-memref-alias-ops.mlir [hunk 2]
```diff
@@ -266,6 +266,127 @@ func.func @fold_vector_transfer_write_with_inner_rank_reduced_subview(
...
+//  CHECK-SAME:    %[[ARG1:[a-zA-Z0-9]+]]: index
+//  CHECK-SAME:    %[[ARG2:[a-zA-Z0-9]+]]: index
+//  CHECK-SAME:    %[[ARG3:[a-zA-Z0-9]+]]: index
+//  CHECK-SAME:    %[[ARG4:[a-zA-Z0-9]+]]: index
+//  CHECK-SAME:    %[[ARG5:[a-zA-Z0-9]+]]: index
+//  CHECK-SAME:    %[[ARG6:[a-zA-Z0-9]+]]: index
+//  CHECK-SAME:    %[[MASK:[a-zA-Z0-9]+]]: vector<4xi1>
+//   CHECK-DAG:    %[[IDX0:.+]] = affine.apply #[[MAP1]]()[%[[ARG1]], %[[ARG5]]]
+//   CHECK-DAG:    %[[IDX1:.+]] = affine.apply #[[MAP1]]()[%[[ARG2]], %[[ARG6]]]
+//       CHECK:    vector.transfer_read %[[ARG0]][%[[IDX0]], %[[IDX1]]], %{{.*}}, %[[MASK]] {{.*}} : memref<?x?xf32
```
Comment (qedawkins) [inline mlir/test/Dialect/MemRef/fold-memref-alias-ops.mlir:293]: Should I keep both tests, or only the rank reducing one?

Comment (nicolasvasilache) [reply mlir/test/Dialect/MemRef/fold-memref-alias-ops.mlir:293]: as far as tests go, the more the better, thanks! :)

**Review by nicolasvasilache — APPROVED** [review]

**Pull request merged**